### PR TITLE
SOF-7207: add numbers to browser timeouts + readme updates

### PR DESCRIPTION
**Pull request by k0stik**

### CODE REVIEW

File: src/js/cypress/Browser.ts [hunk 1]
```diff
@@ -253,28 +253,32 @@ export class Browser extends BaseBrowser {
...
+        delay?: BrowserTimeout | number,
+        timeout?: BrowserTimeout | number,
     ): void;
 
     retry<T = boolean>(
         cb: () => Cypress.Chainable<T>,
         become?: T,
-        delay: BrowserTimeout = "zero",
-        timeout: BrowserTimeout = TimeoutType.md,
+        delay: BrowserTimeout | number = "zero",
```
Comment (k0stik) [inline src/js/cypress/Browser.ts:280]: Numbers here are required in web-app for steps like 
`I see jobs with the following data in {string} status within {string} seconds:`

File: README.md [hunk 2]
```diff
@@ -37,3 +37,125 @@ and JavaScript/TypeScript. In order to set up the pre-commit hooks in the JS/TS
...
+- `getIsVisible`
+- `getIsSelected`
+- `getIsChecked`
+- `getIsExisting`
+
+Browser util methods:
+
+- `retry`
+- `dispatchEvent`
+- `execute`
```
Comment (timurbazhirov) [inline README.md]: We need to clarify which methods to return Cypress Chainable - should be getters only

File: README.md [hunk 3]
```diff
@@ -37,3 +37,125 @@ and JavaScript/TypeScript. In order to set up the pre-commit hooks in the JS/TS
 the `bootstrap.js` script was added. With `pre-commit` installed, running the `bootstrap.js` script is equivalent
 to `husky install`. Note that the hooks are only activated when the package is installed locally (`npm install`)
 and not when installed as a dependency.
+
+## Writing tests
+
+### Common architecture
+
+![alt text](image.png)
```
Comment (timurbazhirov) [inline README.md]: We should explain the isolation of concerns and where cy and browser methods should be used

File: README.md (hunk 2 above)
Comment (k0stik) [reply README.md]: Done

File: README.md (hunk 3 above)
Comment (k0stik) [reply README.md]: Done

File: README.md [hunk 4]
```diff
@@ -37,3 +37,134 @@ and JavaScript/TypeScript. In order to set up the pre-commit hooks in the JS/TS
...
+
+Intermediate level between widgets and test framework (cypress in this case).
+It is available as Cypress window property (`window.browser`) for step definitions and as `this.browser` for any widget and tao.
+Any cypress functions or commands should be used only through the browser object.
+There are 4 types of methods in the browser object:
+
+1. actions
+2. assertions
+3. getters
+4. utils
```
Comment (VsevolodX) [inline README.md]: Maybe let's give a description/definition for each of the method type?

**Review by VsevolodX — COMMENTED** [review]
Having seen examples and worked with tests, it seems clear

File: src/js/cypress/commands/until.ts [hunk 5]
```diff
@@ -69,6 +69,20 @@ Cypress.Commands.add(
     },
 );
 
+Cypress.Commands.add(
+    "thenWithNull",
```
Comment (timurbazhirov) [inline src/js/cypress/commands/until.ts:73]: Let's add a docstring

File: src/js/cypress/commands/until.ts [hunk 6]
```diff
@@ -78,9 +92,13 @@ declare global {
     }
     // eslint-disable-next-line @typescript-eslint/no-namespace
     namespace Cypress {
-        interface Chainable {
+        interface Chainable<Subject> {
```
Comment (timurbazhirov) [inline src/js/cypress/commands/until.ts:113]: Consider adding a docstring for Typescript to read

File: src/js/cypress/commands/until.ts (hunk 5 above)
Comment (k0stik) [reply src/js/cypress/commands/until.ts:73]: Done

File: src/js/cypress/commands/until.ts (hunk 6 above)
Comment (k0stik) [reply src/js/cypress/commands/until.ts:113]: Added docstring

File: README.md (hunk 4 above)
Comment (k0stik) [reply README.md]: Rewrote this with a better explanation

**Review by timurbazhirov — APPROVED** [review]

**Pull request merged**